pathfinder.go

```diff
@@ -0,0 +1,118 @@
+package main
+
+import (
+	"context"
+	"runtime"
+	"sync"
+
+	"golang.org/x/mod/module"
+
+	"github.com/CrowdStrike/perseus/perseusapi"
+)
+
+// newPathFinder initializes and returns a new [pathFinder] instance using the provided Perseus
+// client, maximum depth, and status callback.
+func newPathFinder(c perseusapi.PerseusServiceClient, maxDepth int, status func(string)) pathFinder {
+	return pathFinder{
+		c:        c,
+		maxDepth: maxDepth,
+		status:   status,
+	}
+}
+
+// pathFinder queries the Perseus database to contruct dependency paths of up to maxDepth steps between
+// two modules.
+type pathFinder struct {
+	c        perseusapi.PerseusServiceClient
+	status   func(string)
+	maxDepth int
+
+	sem chan struct{}
+	wg  *sync.WaitGroup
+}
+
+// pathFinderResult defines the result items produced by [pathFinder.findPathsBetween].  Each result
+// contains either a slice of [module.Version] instances representing a path between the specified modules
+// or an error.
+type pathFinderResult struct {
+	path []module.Version
+	err  error
+}
+
+// findPathsBetween repeatedly queries the Perseus server to find one or more dependency paths between
+// the two specified modules.
+func (pf *pathFinder) findPathsBetween(ctx context.Context, from, to module.Version) chan pathFinderResult {
+	// semaphore to limit concurrency to the number of available CPUs
+	n := runtime.NumCPU()
```

Comment:
Since this is more of an IO-bound situation (making calls to the Perseus server) as opposed to a CPU bound situation, do you think limiting the parallel go routines by number of CPUs is ideal? I would think you could be more aggressive than that. WDYT?

Reply:
We probably could but I had to pick something and didn't want overwhelm the user's machine by accident.  We could make it configurable, though.